### PR TITLE
Provide unique ids by checking against already generated ids

### PR DESCRIPTION
**Pull request by midokura-silvia**
IdProvider was not guaranteed to provide an id only once. Added tests.

### CODE REVIEW

File: syft/generic/id_provider.py [hunk 1]
```diff
@@ -13,7 +17,9 @@ class IdProvider:
 
     def __init__(self, given_ids=list()):
         self.given_ids = given_ids
-        self.generated = []
+        self.generated = (
+            []
+        )  # TODO: use sorted container to quickly check whether a value is contained
```
Comment (mari-linhares) [inline syft/generic/id_provider.py]: Can we use a set before merging this to dev?

Comment (iamtrask) [reply syft/generic/id_provider.py]: +1

Comment (midokura-silvia) [reply syft/generic/id_provider.py]: I tried, but it is not as easy as it seems. There is a mechanism to retrieve the list of generated ids via a ResponseSignatureError. And the order of the elements in the list is used to assign ids to return values. So simply changing self.generated to a set() results in errors in the unittests with remote method calls that have more than one return value.

Comment (midokura-silvia) [reply syft/generic/id_provider.py]: The generated ids are now stored in a set.

File: syft/generic/id_provider.py [hunk 2]
```diff
@@ -27,6 +34,44 @@ def pop(self, *args) -> int:
...
+    def set_next_ids(self, given_ids: List):
+        """Sets the next ids returned by the id provider
+
+        Note that the ids are returned in reverse order of the list, as a pop() operation is applied
+
+        Args:
+            given_ids: List, next ids returned by the id provider
+
+        """
+        # TODO: should we check whether these ids conflict with the content of self.generated?
```
Comment (mari-linhares) [inline syft/generic/id_provider.py]: Yes. I think an exception should be thrown.

**Pull request merged**